### PR TITLE
Upgrade to V6

### PR DESCRIPTION
**Pull request by danielmarbach**
This is the first GO to make CustomCheck plugin work with V6. The new checks are async enabled including the ServiceControl backend. Since SCBackend is supposed to be shared infrastructure this probably no longer holds true with the changes introduced in this PR. Open questions that can be addressed in different PRs:
- Do custom checks really make sense, since the silently fail in the background when the SC queues cannot be reached? If the answer is we want to keep them should they retry? What is the implication of a retry since we would retry user code as well?
- Should we move out the samples in this repo?
- Currently we use interface to detect the checks, this suggests anyone can implement these interfaces and their checks will just work. Fact is this plugin assumes that you inherit from the abstract classes provided because without the SC backend wired up there is nothing you can do. I have a trick in my sleeves we could use to make this API more clear.
## Example of new custom check

```
class FailingCustomCheck : CustomCheck
{
    public FailingCustomCheck()
        : base("FailingCustomCheck", "CustomCheck")
    {
    }

    public override Task PerformCheck()
    {
        return ReportFailed("Some reason");
    }
}
```

As you can see it no longer assumes that the work has to be done in the constructor
## Example of a new periodic check

```
class FailingPeriodicCustomCheck : PeriodicCheck
{
    public FailingPeriodicCustomCheck(): base("FailingPeriodicCustomCheck", "PeriodicChecks", TimeSpan.FromSeconds(30))
    {
    }

    public override Task<CheckResult> PerformCheck()
    {
        return CheckResult.Failed("Periodic check failed");
    }
}

class SuccessfulPeriodicCustomCheck : PeriodicCheck
{
    public SuccessfulPeriodicCustomCheck()
        : base("SuccessfulPeriodicCustomCheck", "PeriodicChecks", TimeSpan.FromSeconds(30))
    {
    }

    public override Task<CheckResult> PerformCheck()
    {
        return CheckResult.Pass;
    }
}
```

CheckResult is implicitely convertable to Task<CheckResult> to make the transition easier. With that users don't need to write `Task.FromResult<CheckResult>` all the time if their not doing anything async in their period checks.
## Doco

Coming soon

### CODE REVIEW

**Comment by danielmarbach** [comment]
@indualagarsamy @johnsimons have a look at the early stage of a revamped custom check and periodic check API and how I tried to simplify the code. What do you think?

**Comment by johnsimons** [comment]
it looks good so far

**Comment by danielmarbach** [comment]
@Particular/servicecontrol-maintainers can you review?

File: src/ServiceControl.Plugin.CustomChecks/Internal/ICheck.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace ServiceControl.Plugin.CustomChecks.Internal
+{
+    public interface ICheck
```
Comment (indualagarsamy) [inline src/ServiceControl.Plugin.CustomChecks/Internal/ICheck.cs:3]: Seems like an unnecessary abstraction.

Comment (danielmarbach) [inline src/ServiceControl.Plugin.CustomChecks/Internal/ICheck.cs:3]: I introduced it for the category and the id to cleanup the backend code. I
plan to go towards abstract classes only after the doco. Ok if I
address this later?

**Comment by mikeminutillo** [comment]
I've never been happy with the difference between a `PeriodicCheck` and a `CustomCheck`. Can these be collapsed into a single Custom Check with a `Repeat()` clause or something?

I envisage something like this:

```
public class MyCheck : CustomCheck
{
    public MyCheck() : base( "My Check", "Checks" )
    {
        RepeatEvery(TimeSpan.FromSeconds(30));
    }

    public override Task<CheckResult> PerformCheck()
    {
        return CheckResult.Passed;
    }
}
```

At startup, we'd resolve them all, execute them all and set up timers for those that have a schedule set.

**Comment by mikeminutillo** [comment]
BTW :+1: x :100: on moving the CustomCheck code out of the Constructor.

**Comment by mikeminutillo** [comment]
Actually this would make it easier to have a proper abstraction.

```
public interface ICustomCheck
{
    string Category { get; set; }
    string Id { get; set; }
    TimeSpan? Period { get; set; }
    Task<CheckResult> PerformCheck();
}
```

Then you really could just implement that and have the infrastructure pick it up and fit it into the Custom Checks system.

**Comment by mauroservienti** [comment]
> At startup, we'd resolve them all, execute them all and set up timers for those that have a schedule set.

I like it, :+1:

**Comment by danielmarbach** [comment]
I'll treat the suggestions in a sep. PR.

**Comment by mauroservienti** [comment]
@mikeminutillo does it mean that we can get rid of the 2 interfaces and collapse in one `ICheck` or `ICustomCheck`?

Do we have any requirement in `Pulse` to differentiate results visualization between periodic and startup checks?

**Comment by mauroservienti** [comment]
Are we actually telling to Control/Pulse when the next check will be performed?

**Comment by danielmarbach** [comment]
@mauroservienti yes we can get rid of the two interfaces

Currently pulse doesn't differentiate

**Comment by danielmarbach** [comment]
@mauroservienti 
image

Pulse has the notion of categories. But they are set by the user

**Comment by indualagarsamy** [comment]
I'd keep the Samples in the repo. 
This repo doesn't have an acceptance test. I tried at one point to adding the acceptance test framework, and the effort wasn't trivial. So the easy alternative was to have the sample.

**Comment by mauroservienti** [comment]
@indualagarsamy :+1: until we find a good and easy way to have ATT the involves SC let's keep it simple.

**Pull request merged**